ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/HttpRequestRetry.kt

```diff
@@ -157,8 +156,25 @@ public class HttpRequestRetry internal constructor(configuration: Configuration)
          * Enables retrying a request if an exception is thrown during the [HttpSend] phase
          * and specifies the number of retries.
          */
+        @Deprecated(level = DeprecationLevel.HIDDEN, message = "This will be removed")
         public fun retryOnException(maxRetries: Int = -1) {
-            retryOnExceptionIf(maxRetries) { _, cause -> cause !is CancellationException }
+            retryOnException(maxRetries, false)
+        }
+
+        /**
+         * Enables retrying a request if an exception is thrown during the [HttpSend] phase
+         * and specifies the number of retries.
+         * By default, [HttpRequestTimeoutException] is not retried. Set [retryOnTimeout] to `true` to retry on timeout.
+         * Note, that in this case, [HttpTimeout] plugin should be installed after [HttpRequestRetry].
+         */
+        public fun retryOnException(maxRetries: Int = -1, retryOnTimeout: Boolean = false) {
+            retryOnExceptionIf(maxRetries) { _, cause ->
+                when {
+                    cause is CancellationException -> false
+                    cause is HttpRequestTimeoutException && retryOnTimeout -> true
```

Comment:
Should be `cause is HttpRequestTimeoutException -> retryOnTimeout`? Otherwise if `retryOnTimeout` is false it defaults to the `else` below which is `true`. See comment on tests as well